engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/MethodReferences.scala

```diff
@@ -0,0 +1,201 @@
+
+import scala.annotation.unused
+
+case object MethodReferences extends IRPass {
+
+  /** The type of the metadata object that the pass writes to the IR. */
+  override type Metadata = BindingsMap.Resolution
+
+  /** The type of configuration for the pass. */
+  override type Config = IRPass.Configuration.Default
+
+  /** The passes that this pass depends _directly_ on to run. */
+  override val precursorPasses: Seq[IRPass] =
+    Seq(AliasAnalysis, BindingAnalysis)
+
+  /** The passes that are invalidated by running this pass. */
+  override val invalidatedPasses: Seq[IRPass] = Seq(AliasAnalysis)
+
+  /** Executes the pass on the provided `ir`, and returns a possibly transformed
+    * or annotated version of `ir`.
+    *
+    * @param ir            the Enso IR to process
+    * @param moduleContext a context object that contains the information needed
+    *                      to process a module
+    * @return `ir`, possibly having made transformations or annotations to that
+    *         IR.
+    */
+  override def runModule(
+    ir: IR.Module,
+    moduleContext: ModuleContext
+  ): IR.Module = {
+    val scopeMap = ir.unsafeGetMetadata(
+      BindingAnalysis,
+      "No binding analysis on the module"
+    )
+    val freshNameSupply = moduleContext.freshNameSupply.getOrElse(
+      throw new CompilerError(
+        "No fresh name supply passed to UppercaseNames resolver."
+      )
+    )
+    val new_bindings =
+      ir.bindings.map(processModuleDefinition(_, scopeMap, freshNameSupply))
+    ir.copy(bindings = new_bindings)
+  }
+
+  /** Executes the pass on the provided `ir`, and returns a possibly transformed
+    * or annotated version of `ir` in an inline context.
+    *
+    * @param ir            the Enso IR to process
+    * @param inlineContext a context object that contains the information needed
+    *                      for inline evaluation
+    * @return `ir`, possibly having made transformations or annotations to that
+    *         IR.
+    */
+  override def runExpression(
+    ir: IR.Expression,
+    inlineContext: InlineContext
+  ): IR.Expression = {
+    val scopeMap = inlineContext.module.getIr.unsafeGetMetadata(
+      BindingAnalysis,
+      "No binding analysis on the module"
+    )
+    val freshNameSupply = inlineContext.freshNameSupply.getOrElse(
+      throw new CompilerError(
+        "No fresh name supply passed to UppercaseNames resolver."
+      )
+    )
+    processExpression(ir, scopeMap, freshNameSupply)
+  }
+
+  /** @inheritdoc */
+  override def updateMetadataInDuplicate[T <: IR](
+    @unused sourceIr: T,
+    copyOfIr: T
+  ): T = copyOfIr
+
+  private def processModuleDefinition(
+    definition: IR.Module.Scope.Definition,
+    bindings: BindingsMap,
+    freshNameSupply: FreshNameSupply
+  ): IR.Module.Scope.Definition = {
+    definition match {
+      case asc: IR.Type.Ascription => asc
+      case a =>
+        a.mapExpressions(processExpression(_, bindings, freshNameSupply))
```

Comment:
Off-topic (a bit): As far as I know `mapExpressions` always creates a `copy` - while it should only create a copy if the underlying references changed - e.g. if there is an identity function sent to `mapExpressions`, no copying shall happen at all.

Reply:
That is a valid concern. I think we could short circuit it by checking if the result of the application of the function returns the same object and not copy if that is the case